File: src/main/java/myLang/block/statement/AssignmentContext.java
package myLang.block.statement;

import myLang.block.BlockContext;
import myLang.ContextHandler;
import myLang.response.*;
import myLangParser.MyLangLexer;
import myLangParser.MyLangParser;

public class AssignmentContext implements ContextHandler<MyLangParser.AssignmentContext> {
    @Override
    public Response handler(MyLangParser.AssignmentContext ctx) throws MyLangException {
        String name = ctx.ID().getText();
        Integer num = (Integer) BlockContext.get(name).getValue();
        IntegerResponse response = (IntegerResponse) defaultHandler(ctx, 2);
        switch (ctx.assignmentOperator().start.getType()) {
            case MyLangLexer.ASSIGN:
                return new IntegerResponse(BlockContext.update(name, response.getResponse()));
            case MyLangLexer.ADD_ASSIGN:
                return new IntegerResponse(BlockContext.update(name, num + response.getResponse()));
            case MyLangLexer.SUB_ASSIGN:
                return new IntegerResponse(BlockContext.update(name,num - response.getResponse()));
            case MyLangLexer.MUL_ASSIGN:
                return new IntegerResponse(BlockContext.update(name,num * response.getResponse()));
            case MyLangLexer.DIV_ASSIGN:
                if(response.getResponse() == 0){
                    throw new MyLangException("Деление на 0");
                }
                return new IntegerResponse(BlockContext.update(name,num / response.getResponse()));
            default:
                throw new MyLangException("Операция " + ctx.assignmentOperator().start.getText() + " не поддерживается");
        }
    }
}

File: src/main/kotlin/myLangKotlin/block/statement/AssignmentContext.kt
package myLangKotlin.block.statement

import myLangKotlin.ContextHandler
import myLangKotlin.response.IntegerResponse
import myLangKotlin.response.MyLangException
import myLangKotlin.response.Response
import myLangKotlin.block.BlockContext
import myLangParser.MyLangLexer
import myLangParser.MyLangParser

class AssignmentContext : ContextHandler<MyLangParser.AssignmentContext> {
    @Throws(MyLangException::class)
    override fun handler(ctx: MyLangParser.AssignmentContext): Response<*> {
        val name = ctx.ID().text
        val num = BlockContext[name].value as Int
        val response = defaultHandler(ctx, 2) as IntegerResponse
        when (ctx.assignmentOperator().start.type) {
            MyLangLexer.ASSIGN -> return IntegerResponse(BlockContext.update(name, response.response))
            MyLangLexer.ADD_ASSIGN -> return IntegerResponse(BlockContext.update(name, num + response.response))
            MyLangLexer.SUB_ASSIGN -> return IntegerResponse(BlockContext.update(name, num - response.response))
            MyLangLexer.MUL_ASSIGN -> return IntegerResponse(BlockContext.update(name, num * response.response))
            MyLangLexer.DIV_ASSIGN -> {
                if (response.response == 0) {
                    throw MyLangException("Деление на 0")
                }
                return IntegerResponse(BlockContext.update(name, num / response.response))
            }
            else -> throw MyLangException("Операция " + ctx.assignmentOperator().start.text + " не поддерживается")
        }
    }
}
